File: engine/services/ignoring_service.go
package services

import (
	"context"

	"kouma-engine/models"
	"kouma-engine/utils"

	"go.mongodb.org/mongo-driver/bson"
)

type PlainIgnoring struct {
	PID        string             `json:"pid"`
	CaseName   string             `json:"caseName"`
	Rectangles []models.Rectangle `json:"rectangles"`
}

func GetPlainIgnoring(pid, caseName string) (*PlainIgnoring, error) {
	coll := utils.GetCollection("ignorings")

	var ignoring models.Ignoring
	err := coll.FindOne(context.Background(), bson.M{"pid": pid, "caseName": caseName}).Decode(&ignoring)
	if err != nil {
		return nil, nil
	}

	rectangles := make([]models.Rectangle, len(ignoring.Rectangles))
	for i, r := range ignoring.Rectangles {
		rectangles[i] = models.Rectangle{
			X:      r.X,
			Y:      r.Y,
			Width:  r.Width,
			Height: r.Height,
		}
	}

	return &PlainIgnoring{
		PID:        ignoring.PID,
		CaseName:   ignoring.CaseName,
		Rectangles: rectangles,
	}, nil
}

func GetPlainIgnorings(pid string, caseNames []string) (map[string]*PlainIgnoring, error) {
	coll := utils.GetCollection("ignorings")

	filter := bson.M{
		"pid":      pid,
		"caseName": bson.M{"$in": caseNames},
	}

	cursor, err := coll.Find(context.Background(), filter)
	if err != nil {
		return nil, err
	}
	defer cursor.Close(context.Background())

	result := make(map[string]*PlainIgnoring)
	for cursor.Next(context.Background()) {
		var ignoring models.Ignoring
		if err := cursor.Decode(&ignoring); err != nil {
			return nil, err
		}

		rectangles := make([]models.Rectangle, len(ignoring.Rectangles))
		for i, r := range ignoring.Rectangles {
			rectangles[i] = models.Rectangle{X: r.X, Y: r.Y, Width: r.Width, Height: r.Height}
		}

		result[ignoring.CaseName] = &PlainIgnoring{PID: ignoring.PID, CaseName: ignoring.CaseName, Rectangles: rectangles}
	}

	return result, cursor.Err()
}

File: dashboard/backend/services/ignoring-service.ts
import { Ignoring, IIgnoring, IRectangle } from '../models/ignoring';
import { logger } from '../utils/logger.ts';

class IgnoringService {
    plainIgnoring(ignoring: IIgnoring | null) {
        if (!ignoring) {
            return ignoring;
        }

        return {
            pid: ignoring.pid,
            caseName: ignoring.caseName,
            rectangles: ignoring.rectangles.map((rectangle: IRectangle) => {
                return { x: rectangle.x, y: rectangle.y, width: rectangle.width, height: rectangle.height };
            }),
        };
    }

    async createOrUpdateIgnoring(pid: string, caseName: string, rectangles: IRectangle[]) {
        let ignoring = await this.getIgnoring(pid, caseName);

        if (ignoring) {
            if (rectangles.length) {
                await ignoring.resetRectangles(rectangles);
            } else {
                logger.info(`IGNORING-SERVICE: rectangles is empty, delete current ignoring for pid=${pid}, caseName=${caseName}`);
                await this.deleteIgnoring(pid, caseName);
                ignoring = null;
            }
        } else {
            if (rectangles.length) {
                ignoring = new Ignoring();
                await ignoring.create(pid, caseName, rectangles);
            }
        }

        return this.plainIgnoring(ignoring);
    }

    async getIgnoring(pid: string, caseName: string) {
        return Ignoring.findOne({ pid: pid, caseName: caseName });
    }

    async getPlainIgnoring(pid: string, caseName: string) {
        return this.plainIgnoring(await this.getIgnoring(pid, caseName));
    }

    async deleteIgnoring(pid: string, caseName: string) {
        return Ignoring.deleteMany({ pid: pid, caseName: caseName });
    }

    async cleanProjectIgnoring(pid: string) {
        return Ignoring.deleteMany({ pid: pid });
    }
}

export const ignoringService = new IgnoringService();
